server/aesdsocket.c

```diff
@@ -0,0 +1,318 @@
+    exit(EXIT_FAILURE);
+  }
+
+  if ((argc == 2) && (strcmp(argv[1], "-d") == 0)) {
+    pid_t pid;
+    int i = 0;
+
+    /* create new process If parent, log and exit successfully*/
+    pid = fork();
+    if (pid == -1) {
+      syslog(LOG_ERR, "ERROR: FORK failed");
+      exit(EXIT_FAILURE);
+    } else if (pid != 0) {
+      syslog(LOG_INFO, "INFO- Parent");
+      exit(EXIT_SUCCESS);
+    }
+    // Child process continues here
+    syslog(LOG_INFO, "INFO- Child");
+    // Create a new session for the child process
+    if (setsid() == -1) {
+      syslog(LOG_ERR, "ERROR: SESSION FAIL");
+      exit(EXIT_FAILURE);
+    }
+    // Change the current working directory to the root directory
+    if (chdir("/") == -1) {
+      syslog(LOG_ERR, "ERROR: CHDIR FAIL");
+      exit(EXIT_FAILURE);
+    }
+    // Close all file descriptors except 0, 1, and 2
+    for (i = 0; i < 3; i++) {
+      close(i);
+    }
+    /* redirect fd's 0,1,2 to /dev/null */
+    int fd = open("/dev/null", O_RDWR); /* stdin */
+    dup(0);                             /* stdout */
+    dup(0);                             /* stderror */
+    close(fd);
+
+    /* do its daemon thing... */
+  }
+
+  while (transfer_exit != 1) {
+    // Declares a variable to store the size of the client's address structure.
+    struct sockaddr_in client_addr;
+    socklen_t client_len = sizeof(client_addr);
+    // Accepts an incoming connection, obtaining a new socket file descriptor
+    // specifically for that connection.
+    client_socket_fd =
+        accept(socket_fd, (struct sockaddr *)&client_addr, &client_len);
+
+    if (client_socket_fd == -1) {
+      syslog(LOG_ERR, "ERROR: Failed to ACCEPT SOCKET");
+      error_handler(file_work);
+      exit(EXIT_FAILURE);
+    }
+    // Converts the client's IP address from binary to a human-readable string.
+    // simple inversion of gethostbyname concept.
+    if (NULL == inet_ntop(AF_INET, get_in_addr((struct sockaddr *)&client_addr),
+                          ip_addr, sizeof(ip_addr))) {
+      syslog(LOG_ERR, "ERROR: Failed to OBTAIN IP");
+      error_handler(file_work);
+      exit(EXIT_FAILURE);
+    }
+    syslog(LOG_DEBUG, "Accepted connection from %s", ip_addr);
+    char buf[BUF_SIZE];
+    int packet_receive_complete = 0;
+    // Create file to push data to /var/tmp/aesdsocketdata
+    file_fd = open(DATA_FILE, O_CREAT | O_RDWR | O_APPEND, 0744);
+    if (file_fd == -1) {
+      syslog(LOG_ERR,
+             "ERROR: Failed to  OPEN the FILE at /var/temp/aesdsocketdata");
+      error_handler(Socket_created);
+      exit(EXIT_FAILURE);
+    }
+
+    while (1) {
+      // bzero(buf, BUF_SIZE);
+      memset(buf, 0, BUF_SIZE);
+      int n_received = recv(client_socket_fd, buf, 1024, 0);
+      if (n_received == -1) {
+        syslog(LOG_ERR, "ERROR: Failed to RECEIVE data");
+        error_handler(file_work);
+        exit(EXIT_FAILURE);
+      }
+
+      int n_written = write(file_fd, buf, n_received);
```

Comment:
@aneesh1298 with this design you will write to the file before the packet completes, which may cause problems with future assignments.